File: csyd_102/src/ssfinput.cpp
// Folder Input

// SSFInput.cp

#include "ss.h"
#include "ssfinput.h"

SSFInput::SSFInput(ModList * mList, short h, short v) : SSModule(MT_FInput, mList, h, v)
{
	fNbr = 0;
	ClearExp(&defExp);
	strcpy(desc,"input value");
	DescribeLink(0, "INVALID", "???",0xFFFF,0xFFFF,0xFFFF);
}

void	SSFInput::Copy(SSModule *ss)
{
	SSFInput *sa = (SSFInput *) ss;
	SSModule::Copy(ss);
	CopyExp(&sa->defExp, &defExp);
	fNbr = sa->fNbr;
	strcpy(desc,sa->desc);
}

/*
void SSFInput::Save(FILE* ar)
{
	SSModule::Save(ar);
	WriteFileLine(ar,"FIDEF %s\r",defExp.exp);
	WriteFileLine(ar,"FINN %d\r",fNbr);
	WriteFileLine(ar,"FIND %s\r",desc);
}
*/

void SSFInput::Load(FILE* ar)
{
	char	*p,tbuf[256];
	LoadExp(ar,"FIDEF", &defExp);

	p = parList->GetNextInputLine(ar,"FINN",tbuf);
	sscanf(p,"FINN %d",&fNbr);

	p = parList->GetNextInputLine(ar,"FIND",tbuf);
	sscanf(p,"FIND %s",desc);
}

ssfloat SSFInput::GenerateOutput(SSModule *callingMod)
{
	ssfloat	v;
	this->callingMod = callingMod;
	if (!parList->mods[0]->GetFolderSig(fNbr, &v))
		v = SolveExp(&defExp,callingMod);
  lastRightSample = v;
	return v;
}

void SSFInput::Reset(SSModule *callingMod)
{
	SSModule::Reset(callingMod);
	ResetExp(&defExp,callingMod);
}

#ifdef UI_FEATURES
char* SSFInput::GetLabel()
{
	return desc;
}

Boolean SSFInput::GetOverlay(char *overlay, int* x, int* y)
{
	*x = cellBounds.left+11;
	*y = cellBounds.top+20;
	sprintf(overlay,"%d",fNbr);
	return true;
}
#endif

File: jsyd_102/src/SSFInput.java
import java.io.*;
import java.util.*;

public class SSFInput extends SSModule implements SydConstants
{
  public int           fNbr = 0;
  public String        desc = "input value";
  public ExpRec        defExp = InitExp("");

  public SSFInput(ModList mList)
  {
    super(MT_FInput, mList);
    DescribeLink(0, "INVALID", "???");
  }

  void  Copy(SSModule ss)
  {
    SSFInput sa = (SSFInput) ss;
    super.Copy(ss);
    CopyExp(sa.defExp, defExp);
    fNbr = sa.fNbr;
    desc = sa.desc;
  }

  void Save(PrintWriter ar) throws IOException
  {
    super.Save(ar);
    ar.println("FIDEF " + defExp.exp);
    ar.println("FINN " + fNbr);
    ar.println("FIND " + desc);
  }

  void Load(BufferedReader ar) throws IOException
  {
    defExp = LoadExp(ar,"FIDEF");
    fNbr = Integer.parseInt(getFirstToken(ar,"FINN"));
    desc = LoadString(ar,"FIND"); // This needs to read entire rest of line
  }

  double GenerateOutput(SSModule callingMod)
  {
    double v;
    this.callingMod = callingMod;
    try {
      v = ((SSModule) parList.mods.firstElement()).GetFolderSig(fNbr);
    }
    catch (Exception e)
    {
      v = SolveExp(defExp, callingMod);
    }
    lastRightSample = v;
    return v;
  }

  void Reset(SSModule callingMod)
  {
    super.Reset(callingMod);
    ResetExp(defExp, callingMod);
  }

}

